src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4824,4 +4787,84 @@ private int getClassAccessFlagsRaw() {
     }
 
     private native int getClassAccessFlagsRaw0();
+
+    private Method findMethod(boolean publicOnly, String name, Class<?>... parameterTypes) {
+        PublicMethods.MethodList res = getMethodsRecursive(name, parameterTypes, true, publicOnly);
+        return res == null ? null : getReflectionFactory().copyMethod(res.getMostSpecific());
+    }
+
+    /**
+     * Return the first method that meets the requirements of an application main method.
+     * The method must:
+     * <ul>
+     * <li>be declared this class's hierarchy</li>
+     * <li>have the name "main"</li>
+     * <li>have no arguments or a single argument of type {@code String[]}</li>
+     * <li>have the return type of void</li>
+     * <li>be public, protected or package private</li>
+     * <li>not be abstract</li>
+     * </ul>
+     * Main methods that have a {@code String[]} argument will be chosen over main
+     * methods with no argument.
+     * @apiNote The method returned may be declared in this class, a parent class
+     * or as a default method of an interface that the class or parent class
+     * implements. It is not possible to declare a static main method and instance main
+     * method with the same signature in the same class. (@jls 8.4.2) states that
+     * "It is a compile-time error to declare two methods with override-equivalent
+     * signatures in a class."
+     * @return the candidate main method or null if none found
+     * @throws  SecurityException
+     *          If a security manager, <i>s</i>, is present and any of the
+     *          following conditions is met:
+     *          <ul>
+     *          <li> the caller's class loader is not the same as the
+     *          class loader of this class and invocation of
+     *          {@link SecurityManager#checkPermission
+     *          s.checkPermission} method with
+     *          {@code RuntimePermission("accessDeclaredMembers")}
+     *          denies access to the declared methods within this class
+     *          <li> the caller's class loader is not the same as or an
+     *          ancestor of the class loader for the current class and
+     *          invocation of {@link SecurityManager#checkPackageAccess
+     *          s.checkPackageAccess()} denies access to the package
+     *          of this class
+     *          </ul>
+     * @jls 8.2 Class Members
+     * @jls 8.4 Method Declarations
+     * @jls 8.4.2 Method Signature
+     * @since 22
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.IMPLICIT_CLASSES)
+    @CallerSensitive
+    public Method getMainMethod() {
```

Comment:
This is a new addition in the last few days. It's somewhat surprising as the  launch protocol has never really surfaced in the API, there's nothing to identify the main class for example. I'm in two minds on surfacing this but it does make the java launcher code much simpler.

Class.getMethod/getMethods return public methods, the proposed getMainMethod may return a public or non-public method and may walk the class hierarchy. So I'm not sure about the naming, it feels more like a findMainMethod to emphasize the search or just mainMethod.

The javadoc probably needs to be fleshed out a bit more to specify how the main method is found.  It currently says "be declared this class's hierarchy" (typo) but it has to say more and make it clear that it may return a Method with a declared class that may be a superclass.

"(@jls 8.4.2)", I assume that should be "{@jls 8.4.2}".

The return description says "the candidate main method". This begs the question if there are other candidates, how do I find them?

The SM permission check looks right, it has to be the same as getDeclaredMethod.

Reply:
> [...] it does make the java launcher code much simpler.

Yes. That's a very nice effect of surfacing the launch protocol directyl in `java.lang.Class`

> So I'm not sure about the naming, it feels more like a `findMainMethod` to emphasize the search or just `mainMethod`.

+1 for `findMainMethod`, and also an explicit `Optional<Method>` as the return type.

Reply:
Updated.

Optional is stylizing preference. Since this method is used frequently (almost always) at startup, loading extra classes should probably be avoided. It's not hard to simply `Optional.ofNullable(cls.findMainMethod())`.